### ISSUE TITLE
Not Valid Union Constructor errors from GitHub Schema queries

### ISSUE THREAD

**Issue by tylerjl** [morpheusgraphql/morpheus-graphql#514]
Hi! Thank you so much for working on this library, it's been very useful for me.

Given GitHub's public GraphQL schema, the following query:

```
defineByDocumentFile "assets/github.gql"
  [gql|
    query GetTags ($owner: String!, $repo: String!)
      {
        repository(owner: $owner, name: $repo) {
          refs(refPrefix: "refs/tags/", first: 100) {
            pageInfo {
              endCursor
              hasNextPage
            }
            edges {
              cursor
              node {
                name
                target {
                  __typename
                  ... on Tag {
                    tagger {
                      date
                    }
                  }
                }
              }
            }
          }
        }
      }
  |]
```

Results in the following error from the morpheus-graphql-client library:

    Error in $.data.repository.refs.edges[2].node.target: ("Commit",fromList [("__typename",String "Commit")]) is Not Valid Union Constructor

Any ideas? Please let me know if there's any other debugging information I could provide that would be helpful.

**Comment by tylerjl** [comment]
Forgot to mention: I can observe this behavior when viewing a repo like this one (`morpheusgraphql/morpheus-graphql`)  but not one like `tylerjl/overdue`. The raw responses from the GitHub GraphQL explorer for each repo, respectively, look like:

```
...
        "edges": [
          {
            "cursor": "MQ",
            "node": {
              "name": "0.0.1",
              "target": {
                "__typename": "Commit"
              }
            }
          },
...
```

and

```
...
        "edges": [
          {
            "cursor": "MQ",
            "node": {
              "name": "v1.0.0",
              "target": {
                "__typename": "Tag",
                "tagger": {
                  "date": "2016-05-10T21:58:08.000-04:00"
                }
              }
            }
          },
...
```

So it seems like an issue with the what sort of `__typename` is returned. Logically, the error makes sense (`Commit`might not have the same fields as a `Tag` but both implement `GitObject`), but I don't know enough to go much further than that.

**Comment by nalchevanidze** [comment]
thanks @tylerjl  if you provide `assets/github.gql` would be good start :)

**Comment by tylerjl** [comment]
Thanks @nalchevanidze - I'm using the unmodified GitHub public schema, [which is the one here](https://docs.github.com/public/schema.docs.graphql).

**Comment by nalchevanidze** [comment]
@tylerjl this issue showed me another problem, parser was too slow. 

it took around 20-25 to parse this schema https://docs.github.com/public/schema.docs.graphql

now with #533. it is around 15-20x faster. takes around 1.7-2 sec

**Comment by nalchevanidze** [comment]
@tylerjl is fixed in #535

**Comment by tylerjl** [comment]
Thank you so much for the responsive fix!
